File: leetcode/字符串匹配/17.13.py
# 17.13. 恢复空格 - 中等
# https://leetcode-cn.com/problems/re-space-lcci/


# 2. 尝试建立完全树索引 - wip
class Solution:
    def respace(self, dictionary: List[str], sentence: str) -> int:
        idxDct = self.buildIndexDict(dictionary)

        N = len(sentence)
        max_reco = 0
        i = 0
        memo = {}

        def recurse(curr_sen: str, curr_reco: int) -> None:
            nonlocal max_reco, memo
            if not curr_sen:
                return
            if curr_sen in memo:
                return
            left, first = curr_sen[1:], curr_sen[0]

            recurse(left, curr_reco)
            memo[left] = 1

            p = idxDct
            while first in p:  # 首字符匹配
                left, first = first[1:], first[0]
                curr_reco += 1
                max_reco = max(max_reco, curr_reco)
                p = p[first]

            recurse(left, curr_reco)
            memo[left] = 1

        recurse(sentence, 0)
        return N - max_reco

    # 构建字典索引树
    def buildIndexDict(self, dictionary: List[str]):
        # https://blog.csdn.net/bencodeben/java/article/details/47955381
        def deepMerge(dict1, dict2) -> None:
            for key in dict2.keys():
                if key not in dict1.keys():
                    dict1[key] = dict2[key]
                else:
                    deepMerge(dict1[key], dict2[key])

        def recurse(word: str, currDct) -> None:
            if not word:
                return
            first, left = word[0], word[1:]
            nextDct = {}
            recurse(left, nextDct)
            if first in currDct:
                deepMerge(currDct[first], nextDct)
            else:
                currDct[first] = nextDct

        dct = {}
        for word in dictionary:
            recurse(word, dct)
        return dct


# 1. 暴力递归+记忆化+首字符索引(lv1) - 超时
class Solution:
    def respace(self, dictionary: List[str], sentence: str) -> int:
        indexes = {}  # 首字符 索引 level1
        for word in dictionary:
            first = word[0]
            indexes[first] = indexes[first] + \
                [word] if first in indexes else [word]

        memo = {}  # 记忆化
        reco = 0

        def recurse(s: str, acc: int) -> None:
            nonlocal reco, memo
            if not s or (s in memo and memo[s] >= acc):
                return
            memo[s] = acc
            first = s[0]
            if first in indexes:  # 首字符匹配 尝试单词匹配
                for word in indexes[first]:
                    wlen = len(word)
                    if s[0: wlen] == word:  # 单词匹配
                        # print(["matched", word, s[wlen:]])
                        reco = max(reco, acc + wlen)
                        recurse(s[wlen:], acc + wlen)  # 选择匹配单词 跳过单词
                        if not s[wlen:]:
                            # print(["break", s[wlen:], acc + wlen])
                            break
                    recurse(s[1:], acc)  # 选择不匹配单词 下一位字符
            else:
                recurse(s[1:], acc)  # 首字符不匹配 下一位字符

        recurse(sentence, 0)
        return len(sentence) - reco

File: leetcode/字符串匹配/17.13.go
// 面试题 17.13. 恢复空格 - 中等
// https://leetcode-cn.com/problems/re-space-lcci/
// https://leetcode-cn.com/problems/re-space-lcci/solution/hui-fu-kong-ge-by-leetcode-solution/


// 2. 方法二：字符串哈希
// 字符串哈希：可参考「1392. 最长快乐前缀」官方题解中的「背景知识」。
// 我们使用字典树的目的是查找某一个串 s 是否在一个串的集合 S 当中，并且当我们知道 s 是否在 S 中之后，可以快速的知道在 s 后添加某一个新的字母得到的新串 s' 是否在 S 中，这个转移的过程是 O(1) 的。
// 执行用时： 52 ms , 在所有 Go 提交中击败了 87.50% 的用户
// 内存消耗： 4.8 MB , 在所有 Go 提交中击败了 100.00% 的用户
const (
    P = math.MaxInt32
    BASE = 41
)

func respace(dictionary []string, sentence string) int {
    hashValues := map[int]bool{}
    for _, word := range dictionary {
        hashValues[getHash(word)] = true
    }
    f := make([]int, len(sentence) + 1)
    for i := 1; i < len(f); i++ {
        f[i] = len(sentence)
    }
    for i := 1; i <= len(sentence); i++ {
        f[i] = f[i-1] + 1
        hashValue := 0
        for j := i; j >= 1; j-- {
            t := int(sentence[j-1] - 'a') + 1
            hashValue = (hashValue * BASE + t) % P
            if hashValues[hashValue] {
                f[i] = min(f[i], f[j-1])
            }
        }
    }
    return f[len(sentence)]
}

func getHash(s string) int {
    hashValue := 0
    for i := len(s) - 1; i >= 0; i-- {
        hashValue = (hashValue * BASE + int(s[i] - 'a') + 1) % P
    }
    return hashValue
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}


// 1. 方法一：Trie + 动态规划
// 字典树 Trie
// 定义 dp[i] 表示考虑前 i 个字符最少的未识别的字符数量，从前往后计算 dp 值。
// 执行用时： 84 ms , 在所有 Go 提交中击败了 37.50% 的用户
// 内存消耗： 52 MB , 在所有 Go 提交中击败了 100.00% 的用户
func respace(dictionary []string, sentence string) int {
    n, inf := len(sentence), 0x3f3f3f3f
    root := &Trie{next: [26]*Trie{}}
    for _, word := range dictionary {
        root.insert(word)
    }
    dp := make([]int, n + 1)
    for i := 1; i < len(dp); i++ {
        dp[i] = inf
    }
    for i := 1; i <= n; i++ {
        dp[i] = dp[i-1] + 1
        curPos := root
        for j := i; j >= 1; j-- {
            t := int(sentence[j-1] - 'a')
            if curPos.next[t] == nil {
                break
            } else if curPos.next[t].isEnd {
                dp[i] = min(dp[i], dp[j-1])
            }
            if dp[i] == 0 {
                break
            }
            curPos = curPos.next[t]
        }
    }
    return dp[n]
}

type Trie struct {
    next [26]*Trie
    isEnd bool
}

func (this *Trie) insert(s string) {
    curPos := this
    for i := len(s) - 1; i >= 0; i-- {
        t := int(s[i] - 'a')
        if curPos.next[t] == nil {
            curPos.next[t] = &Trie{next: [26]*Trie{}}
        }
        curPos = curPos.next[t]
    }
    curPos.isEnd = true
}

func min(x, y int) int {
    if x < y {
				return x
		}
		return y
}
